pkg/operator/startupmonitorreadiness/readiness_checks.go

```diff
@@ -0,0 +1,259 @@
+		//	TODO: watch /var/log/kube-apiserver/termination.log for more than one start-up attempt. Set Reason=CrashLooping if more than one is found and the monitor times out.
+
+		// doRevisionCheck in "lazy" mode
+		func() (bool, string, string) {
+			return doRevisionCheck(ctx, ch.kubeClient.CoreV1().Pods(operatorclient.TargetNamespace), revision, false)
+		},
+
+		// checks etcd health condition
+		func() (bool, string, string) { return doETCDHealthCheck(ctx, ch.client, ch.baseRawURL) },
+
+		// checks overall Kube API health condition
+		func() (bool, string, string) { return doKubeHealthCheck(ctx, ch.client, ch.baseRawURL) },
+
+		// check overall Kube API readiness condition
+		func() (bool, string, string) {
+			return doKubeReadyCheck(ctx, ch.client, ch.baseRawURL, 3, 5*time.Second)
+		},
+
+		// doRevisionCheck in "strict" mode
+		func() (bool, string, string) {
+			return doRevisionCheck(ctx, ch.kubeClient.CoreV1().Pods(operatorclient.TargetNamespace), revision, true)
+		},
+
+		//	TODO: get the mirror pod from the localhost kube-apiserver
+		//	TODO: checking status.ready. Set Reason=NotReady if this is red and the monitor times out.
+
+	} {
+		select {
+		case <-ctx.Done():
+			return false, "", "", ctx.Err()
+		default:
+		}
+
+		if ready, reason, message := checkFn(); !ready {
+			return ready, reason, message, nil
+		}
+	}
+
+	// at this point Kube API is ready!
+	return true, "", "", nil
+}
+
+// doRevisionCheck checks if the kas pod is running at the expected revision
+func doRevisionCheck(ctx context.Context, podClient corev1client.PodInterface, monitorRevision int, strictMode bool) (bool, string, string) {
+	apiServerPods, err := podClient.List(ctx, metav1.ListOptions{LabelSelector: "apiserver=true"})
+	if err != nil {
+		return !strictMode, "NotReadyError", fmt.Sprintf("falied to get Kube API server pod due to %v", err)
+	}
+	if len(apiServerPods.Items) == 0 {
+		return !strictMode, "NotReady", "waiting for Kube API server pod to show up"
+	}
+	if len(apiServerPods.Items) != 1 {
+		return !strictMode, "NotReady", fmt.Sprintf("unexpected number of Kube API server pods %d, expected only one pod", len(apiServerPods.Items))
+	}
+
+	kasPod := apiServerPods.Items[0]
+	revisionString, found := kasPod.Labels["revision"]
+	if !found {
+		return false, "NotReady", fmt.Sprintf("pod %s doesn't have revision label", kasPod.Name)
+	}
+	if len(revisionString) == 0 {
+		return false, "NotReady", fmt.Sprintf("empty revision label on %s pod", kasPod.Name)
+	}
+	revision, err := strconv.Atoi(revisionString)
+	if err != nil || revision < 0 {
+		return false, "NotReady", fmt.Sprintf("invalid revision label on pod %s: %q", kasPod.Name, revisionString)
+	}
+
+	if revision != monitorRevision {
+		return false, "NotReady", fmt.Sprintf("the running Kube API is at unexpected revision %d, expected %d", revision, monitorRevision)
```

Comment:
use better reasons for all of these. NotReady should be reserved for the time we see the pod in the API, but it is not ready yet.

Reply:
WrongRevision